tests/jetty-jmh/src/main/java/org/eclipse/jetty/server/jmh/DeflaterPoolBenchmark.java

```diff
@@ -92,13 +92,14 @@ public static void stopTrial() throws Exception
     @SuppressWarnings("deprecation")
     public long testPool() throws Exception
     {
-        Deflater deflater = _pool.acquire();
+        DeflaterPool.Entry entry = _pool.acquire();
```

Comment:
... and the answer to the 'how fast/slow' question is???

Reply:
In this benchmark new pool is performing about 50% slower compared to before the changes, I am trying to figure out why.

Reply:
```
Benchmark (Old Implementation)                                         (poolType)   Mode  Cnt       Score        Error   Units
DeflaterPoolBenchmark.testPool                                            NO_POOL  thrpt   10  132272.104 ±  70177.178   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                             NO_POOL  thrpt   10      14.751 ±      0.742  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm                        NO_POOL  thrpt   10     200.392 ±      0.195    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space                    NO_POOL  thrpt   10      18.813 ±     60.160  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm               NO_POOL  thrpt   10     267.667 ±    859.382    B/op
DeflaterPoolBenchmark.testPool:·gc.count                                  NO_POOL  thrpt   10       2.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                                   NO_POOL  thrpt   10       2.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_10  thrpt   10  239372.722 ±  76227.954   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_10  thrpt   10      20.725 ±      1.682  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_10  thrpt   10     125.168 ±      7.707    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_10  thrpt   10      25.633 ±     63.035  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_10  thrpt   10     154.377 ±    381.899    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_10  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_10  thrpt   10       7.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_20  thrpt   10  419346.829 ± 183133.367   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_20  thrpt   10      25.289 ±      2.318  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_20  thrpt   10     107.170 ±      6.393    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_20  thrpt   10      33.894 ±     66.218  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_20  thrpt   10     132.195 ±    259.559    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_20  thrpt   10       4.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_20  thrpt   10       6.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_50  thrpt   10  591581.732 ± 114335.926   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_50  thrpt   10      31.112 ±      6.459  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_50  thrpt   10      89.393 ±      1.233    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_50  thrpt   10      37.886 ±     74.219  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_50  thrpt   10     104.904 ±    206.682    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Survivor_Space       DEFLATER_POOL_50  thrpt   10       0.001 ±      0.005  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Survivor_Space.norm  DEFLATER_POOL_50  thrpt   10       0.003 ±      0.016    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_50  thrpt   10       4.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_50  thrpt   10       6.000                   ms


Benchmark (New Implementation)                                        (poolType)   Mode  Cnt       Score        Error   Units
DeflaterPoolBenchmark.testPool                                            NO_POOL  thrpt   10    5247.848 ±   1251.467   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                             NO_POOL  thrpt   10       0.865 ±      0.136  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm                        NO_POOL  thrpt   10     206.481 ±      3.216    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space                    NO_POOL  thrpt   10       1.667 ±      7.969  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm               NO_POOL  thrpt   10     466.904 ±   2232.227    B/op
DeflaterPoolBenchmark.testPool:·gc.count                                  NO_POOL  thrpt   10       1.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                                   NO_POOL  thrpt   10      57.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_10  thrpt   10   92061.829 ±  65986.254   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_10  thrpt   10       9.736 ±     19.304  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_10  thrpt   10     174.152 ±    367.960    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_10  thrpt   10      13.174 ±     42.006  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_10  thrpt   10     247.383 ±    788.571    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Survivor_Space       DEFLATER_POOL_10  thrpt   10       0.119 ±      0.570  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Survivor_Space.norm  DEFLATER_POOL_10  thrpt   10       2.262 ±     10.814    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_10  thrpt   10       2.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_10  thrpt   10      50.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_20  thrpt   10  183197.471 ± 237236.784   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_20  thrpt   10       5.404 ±      1.575  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_20  thrpt   10      74.410 ±      3.782    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_20  thrpt   10       7.991 ±     38.203  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_20  thrpt   10     113.470 ±    542.492    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_20  thrpt   10       1.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_20  thrpt   10      33.000                   ms
DeflaterPoolBenchmark.testPool                                   DEFLATER_POOL_50  thrpt   10  312313.037 ± 281479.025   ops/s
DeflaterPoolBenchmark.testPool:·gc.alloc.rate                    DEFLATER_POOL_50  thrpt   10      12.045 ±      3.563  MB/sec
DeflaterPoolBenchmark.testPool:·gc.alloc.rate.norm               DEFLATER_POOL_50  thrpt   10      68.236 ±      1.375    B/op
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_50  thrpt   10      15.183 ±     48.426  MB/sec
DeflaterPoolBenchmark.testPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_50  thrpt   10      92.631 ±    295.256    B/op
DeflaterPoolBenchmark.testPool:·gc.count                         DEFLATER_POOL_50  thrpt   10       2.000               counts
DeflaterPoolBenchmark.testPool:·gc.time                          DEFLATER_POOL_50  thrpt   10      32.000                   ms
```

Reply:
Try Thread ID mode for the pool.
How many threads are you using in the benchmark?
I also wonder if the threadlocal cache is worth it in this case??

Reply:
It seems that if the capacity of the pool is lower than the number of threads using it then the old pool is faster, but if the capacity of the pool is larger than the number of deflaters needed then the new pool can be around 20% faster.

These results had 100 threads contending on the pool, but only tested pool sizes of 0,10,20,50 which are all below the number of deflaters needed. It is still quite a large performance reduction if the pool is reaching capacity.

Reply:
@lachlan-roberts Why are the NO_POOL results so different?  Surely with no pool the results should be about the same?  Also the B/ops for NO_POOL are a long way apart, when they should be dominated by the size of the inflater/deflater

When running the benchmark with the old pool in unlimited mode, how many inflaters/deflaters are actually created? Close to the number of threads?

Note that to be fair, I think the benchmark test should always consume some CPU without an inflater/deflater held, then acquire one, then consume some more CPU before releasing.

Reply:
> Why are the NO_POOL results so different?

The `NO_POOL` results are from a bug were we didn't end the deflater if it wasn't pooled. I fixed it  in 5861709cdb75ed12030f3207f12169dcff59a1f0, and now the results are showing pretty equal values for both on the `NO_POOL` setting.

> When running the benchmark with the old pool in unlimited mode, how many inflaters/deflaters are actually created? Close to the number of threads?

In both cases exactly 100 deflaters are created, same as the number of threads.

Reply:
Can you give latest benchmark results with the NO_POOL bug fixed.  Also with scaling to size of the pool == # of threads.

If previously we defaulted to unlimited pool and it always grew to the number of threads, then I think initializing the compressions pools capacity using the threadpool size is not a bad thing to do.

Reply:
```
Benchmark (New Implementation)                                                     (poolType)   Mode  Cnt       Score        Error   Units
DeflaterPoolBenchmark.testNewPool                                                     NO_POOL  thrpt   10  115818.366 ±  69230.105   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                                      NO_POOL  thrpt   10      12.458 ±      2.487  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm                                 NO_POOL  thrpt   10     200.510 ±      0.440    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space                             NO_POOL  thrpt   10       9.142 ±     43.707  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm                        NO_POOL  thrpt   10     150.607 ±    720.040    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                                           NO_POOL  thrpt   10       1.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                                            NO_POOL  thrpt   10       4.000                   ms
DeflaterPoolBenchmark.testNewPool                                            DEFLATER_POOL_10  thrpt   10  159444.955 ±  27752.686   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                             DEFLATER_POOL_10  thrpt   10      16.310 ±      3.020  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm                        DEFLATER_POOL_10  thrpt   10     142.825 ±     10.737    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_10  thrpt   10      22.320 ±     56.242  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_10  thrpt   10     216.805 ±    531.429    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                                  DEFLATER_POOL_10  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                                   DEFLATER_POOL_10  thrpt   10       4.000                   ms
DeflaterPoolBenchmark.testNewPool                                            DEFLATER_POOL_20  thrpt   10  236983.110 ±  79024.092   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                             DEFLATER_POOL_20  thrpt   10      17.561 ±      3.229  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm                        DEFLATER_POOL_20  thrpt   10     118.167 ±     10.821    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_20  thrpt   10      17.427 ±     55.552  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_20  thrpt   10     119.816 ±    381.923    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                                  DEFLATER_POOL_20  thrpt   10       2.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                                   DEFLATER_POOL_20  thrpt   10       5.000                   ms
DeflaterPoolBenchmark.testNewPool                                            DEFLATER_POOL_50  thrpt   10  386939.698 ±  81445.930   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                             DEFLATER_POOL_50  thrpt   10      20.591 ±      3.800  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm                        DEFLATER_POOL_50  thrpt   10      85.150 ±      3.677    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_50  thrpt   10      26.638 ±     64.892  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_50  thrpt   10     108.538 ±    265.409    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                                  DEFLATER_POOL_50  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                                   DEFLATER_POOL_50  thrpt   10       7.000                   ms
DeflaterPoolBenchmark.testNewPool                                   DEFLATER_POOL_NUM_THREADS  thrpt   10  612596.657 ±  68120.977   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                    DEFLATER_POOL_NUM_THREADS  thrpt   10      25.557 ±      5.133  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm               DEFLATER_POOL_NUM_THREADS  thrpt   10      64.099 ±      0.103    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_NUM_THREADS  thrpt   10      28.000 ±     68.162  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_NUM_THREADS  thrpt   10      68.634 ±    167.115    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                         DEFLATER_POOL_NUM_THREADS  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                          DEFLATER_POOL_NUM_THREADS  thrpt   10       5.000                   ms
DeflaterPoolBenchmark.testNewPool                                       DEFLATER_POOL_DEFAULT  thrpt   10  621950.213 ±  29432.365   ops/s
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate                        DEFLATER_POOL_DEFAULT  thrpt   10      25.693 ±      4.930  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.alloc.rate.norm                   DEFLATER_POOL_DEFAULT  thrpt   10      64.097 ±      0.098    B/op
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space               DEFLATER_POOL_DEFAULT  thrpt   10      27.766 ±     67.688  MB/sec
DeflaterPoolBenchmark.testNewPool:·gc.churn.G1_Eden_Space.norm          DEFLATER_POOL_DEFAULT  thrpt   10      66.178 ±    161.544    B/op
DeflaterPoolBenchmark.testNewPool:·gc.count                             DEFLATER_POOL_DEFAULT  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testNewPool:·gc.time                              DEFLATER_POOL_DEFAULT  thrpt   10       5.000                   ms


Benchmark (Old Implementation)                                                     (poolType)   Mode  Cnt       Score        Error   Units
DeflaterPoolBenchmark.testOldPool                                                     NO_POOL  thrpt   10  128709.448 ±  44535.107   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                                      NO_POOL  thrpt   10      13.929 ±      2.576  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm                                 NO_POOL  thrpt   10     200.435 ±      0.378    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space                             NO_POOL  thrpt   10       7.849 ±     37.525  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm                        NO_POOL  thrpt   10     108.813 ±    520.226    B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                                           NO_POOL  thrpt   10       1.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                                            NO_POOL  thrpt   10       2.000                   ms
DeflaterPoolBenchmark.testOldPool                                            DEFLATER_POOL_10  thrpt   10  267759.422 ± 120516.326   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                             DEFLATER_POOL_10  thrpt   10      19.599 ±      3.774  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm                        DEFLATER_POOL_10  thrpt   10     125.553 ±      6.286    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_10  thrpt   10      22.134 ±     55.318  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_10  thrpt   10     152.248 ±    371.073    B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                                  DEFLATER_POOL_10  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                                   DEFLATER_POOL_10  thrpt   10       9.000                   ms
DeflaterPoolBenchmark.testOldPool                                            DEFLATER_POOL_20  thrpt   10  324418.348 ±  65355.235   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                             DEFLATER_POOL_20  thrpt   10      22.665 ±      3.131  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm                        DEFLATER_POOL_20  thrpt   10     108.269 ±      6.573    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_20  thrpt   10      22.584 ±     57.177  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_20  thrpt   10     107.425 ±    264.003    B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                                  DEFLATER_POOL_20  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                                   DEFLATER_POOL_20  thrpt   10       4.000                   ms
DeflaterPoolBenchmark.testOldPool                                            DEFLATER_POOL_50  thrpt   10  497624.910 ±  50606.113   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                             DEFLATER_POOL_50  thrpt   10      28.294 ±      5.274  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm                        DEFLATER_POOL_50  thrpt   10      89.701 ±      1.276    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space                    DEFLATER_POOL_50  thrpt   10      28.074 ±     68.346  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm               DEFLATER_POOL_50  thrpt   10      87.301 ±    213.483    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Survivor_Space                DEFLATER_POOL_50  thrpt   10      ≈ 10⁻⁴               MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Survivor_Space.norm           DEFLATER_POOL_50  thrpt   10      ≈ 10⁻⁴                 B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                                  DEFLATER_POOL_50  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                                   DEFLATER_POOL_50  thrpt   10       9.000                   ms
DeflaterPoolBenchmark.testOldPool                                   DEFLATER_POOL_NUM_THREADS  thrpt   10  506060.457 ±  96406.859   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                    DEFLATER_POOL_NUM_THREADS  thrpt   10      24.780 ±      4.503  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm               DEFLATER_POOL_NUM_THREADS  thrpt   10      88.128 ±      0.091    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space           DEFLATER_POOL_NUM_THREADS  thrpt   10      27.832 ±     67.884  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm      DEFLATER_POOL_NUM_THREADS  thrpt   10      98.382 ±    241.545    B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                         DEFLATER_POOL_NUM_THREADS  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                          DEFLATER_POOL_NUM_THREADS  thrpt   10       4.000                   ms
DeflaterPoolBenchmark.testOldPool                                       DEFLATER_POOL_DEFAULT  thrpt   10  496070.033 ±  72193.902   ops/s
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate                        DEFLATER_POOL_DEFAULT  thrpt   10      25.327 ±      5.885  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.alloc.rate.norm                   DEFLATER_POOL_DEFAULT  thrpt   10      88.131 ±      0.115    B/op
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space               DEFLATER_POOL_DEFAULT  thrpt   10      28.035 ±     68.337  MB/sec
DeflaterPoolBenchmark.testOldPool:·gc.churn.G1_Eden_Space.norm          DEFLATER_POOL_DEFAULT  thrpt   10      94.490 ±    230.555    B/op
DeflaterPoolBenchmark.testOldPool:·gc.count                             DEFLATER_POOL_DEFAULT  thrpt   10       3.000               counts
DeflaterPoolBenchmark.testOldPool:·gc.time                              DEFLATER_POOL_DEFAULT  thrpt   10       5.000                   ms
```

Reply:
@lachlan-roberts So given those results, what is your recommendation?
Have you tried the thread_id mode for the pool?

@lorban any comments?

Reply:
@gregw I did try the thread_id mode earlier and it seemed to decrease performance further.
And I can't seem to get results which are that different using the `Blackhole.consumeCPU()`, if I keep increasing the values the difference between the two pools is reduced but that is expected.

Reply:
On my machine, the old implementation tops up at ~1M tps with ~60% cpu utilization while the new one reaches ~2M with ~99% cpu utilization, so I suspect a benchmarking problem. 

My theory is that @lachlan-roberts' machine is spending so much time running the Deflater that most of the time all the entries of the pool are found busy in the new implementation, while there's enough contention on the pool itself in the old implementation that deflaters are given more time to finish their job and to be back to the pool. And since creating a new Deflater involves creating a Phantom Ref (which means contending on a single lock) this may explain this. I'd need to investigate on his machine, or get some JFR reports to assert this.

Since the pool is returning an unpooled `Deflater` when it can't find a pooled one, it would be very useful to make the benchmark report how often this happens since this is an expensive operation. Counting that with a `java.util.concurrent.atomic.LongAdder` and exposing the counter via the `CompressionPool` could shed some light on this unexpected behavior.